### PR TITLE
Local Teleporters

### PR DESCRIPTION
**Pull request by Geevies**
* Added a local teleporter circuiboard to the circuit imprinter, a machine that is capable of teleporting matter safely over small distances.

### CODE REVIEW

**Comment by Geevies** [comment]
!review

File: code/modules/mining/machine_local_teleporter.dm [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	if(linked_teleporter && !(stat & NOPOWER) && !(stat & NOPOWER))
+		add_overlay(connected_overlay)
+
+/obj/machinery/mineral/local_teleporter/Crossed(atom/movable/O)
+	if((stat & NOPOWER) || (stat & NOPOWER))
+		return
+	if(ismob(O))
+		to_chat(O, SPAN_NOTICE("You step into \the [src] and get teleported to its linked teleporter."))
+	use_power(active_power_usage)
+	O.forceMove(get_turf(linked_teleporter))
```
Comment (mikomyazaki) [inline code/modules/mining/machine_local_teleporter.dm:36]: Should we check if something has been moved onto the other teleporter somehow, maybe when the power was off, that should block teleportation?

File: code/modules/mining/machine_local_teleporter.dm [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	if(!ismob(O)) // to get anything that lands on us off the teleporter
+		var/list/step_dirs = cardinal
+		step_dirs -= reverse_dir[dir]
+		step(O, pick(step_dirs), 1)
+	spark(O, 3, alldirs)
```
Comment (mikomyazaki) [inline code/modules/mining/machine_local_teleporter.dm:41]: There should be a Destroy proc override that makes its linked teleporter lose its connected overlay and linkedteleporter.

Also should these things be constructable, configurable?

File: code/modules/mining/machine_local_teleporter.dm [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	var/connected_overlay
+
+/obj/machinery/mineral/local_teleporter/Initialize(mapload)
+	..()
+	connected_overlay = icon(icon, "[initial(icon_state)]-connected")
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/mineral/local_teleporter/LateInitialize()
+	var/turf/T = get_step(src, dir)
+	linked_teleporter = locate() in get_step(T, dir)
```
Comment (mikomyazaki) [inline code/modules/mining/machine_local_teleporter.dm:20]: You could make this machine more versatile for other mappers in future by giving it a var/teleport_range and something like this:

```suggestion
	linked_teleporter = locate() in orange(teleport_range, src)
```

File: code/modules/mining/machine_local_teleporter.dm (hunk 2 above)
Comment (Geevies) [reply code/modules/mining/machine_local_teleporter.dm]: construction can probably be added later, yah

File: code/modules/mining/machine_local_teleporter.dm (hunk 3 above)
Comment (Geevies) [reply code/modules/mining/machine_local_teleporter.dm:20]: currently this is configured to only work in a straight line, so having it be orange wouldn't really work or be desirable

File: code/modules/mining/machine_local_teleporter.dm (hunk 1 above)
Comment (Geevies) [reply code/modules/mining/machine_local_teleporter.dm:39]: naw, blocking teleportation would be more annoying than anything else, imo

**Comment by fernerr** [comment]
Dunno how I feel about this, trivializes teleportation as a concept to cut off a little bit of time from using the airlock, when the crux of the issue with airlocks being slow is because the vents take far too long to suck in air compared to delivering it.

Its positioning also feels out of place, and I know we joke about workplace hazards, but this is a massive one.
One wrong step from trying to take a seat at the desk could have you end up in a vaccum.

**Comment by Doxxmedearly** [comment]
Could be something build and installed by science. Further connects the two departments' co-dependence.

**Comment by Sindorman** [comment]
> I am not in favor of it being able to teleport people as that trivializes the airlocks and will lead to accidents where someone steps on it by mistake.

I need this for mining mulebot update to deliver ore boxes into refinery

**Comment by Arrow768** [comment]
> > I am not in favor of it being able to teleport people as that trivializes the airlocks and will lead to accidents where someone steps on it by mistake.
> 
> I need this for mining mulebot update to deliver ore boxes into refinery

This seems like a very bandaid solution.
Imho it would be better to use some sort of automatic unloading station that can pass w/e the mulebots transport (most likely ore if you want them to use the mining airlocks) into a disposals system which leads to the smelter.

**Comment by NonQueueingMatt** [comment]
Stale.

**Pull request closed without merging**